model/jpa/src/main/java/org/jboss/aerogear/unifiedpush/jpa/dao/impl/JPABaseDao.java

```diff
@@ -43,20 +43,20 @@ public void setEntityManager(EntityManager entityManager) {
 
 
     protected TypedQuery<T> createQuery(String jpql) {
-        return entityManager.createQuery(jpql, getType());
+        return entityManager.createQuery(jpql, getType()).setHint("org.hibernate.cacheable", true);
```

Comment:
Does it make sense to cache all the queries?

Reply:
@gunnarmorling mind taking a look ? 

I think, in theory caching queries does seem reasonable

Reply:
I'd be cautious about enabling query caching in such common base class and query method. You are likely better off to enable this for dedicated entity types and queries only (which are unlikely to change often).

Also have you enabled the second-level cache before (query caching will only cache entity ids)? Check out the section on query caching [here](http://www.baeldung.com/hibernate-second-level-cache), it has some good info.

Overall, I think you first should analyse which queries are run very often and then make use of query caching more selectivel.y

Reply:
@gunnarmorling Yes, second level cache has been enabled.
Since we don't have queries that return big amount of data (if we would have, if would be a bug), I was thinking that:
1) If the query is seldom used, it will be just a few bytes and will be evicted as soon as the cache expires. Since it is just a few bytes, it shouldn't be an issue
2) If the query is used often, than it is good for that to be cached
On the other side, if I choose myself which query to cache, I could miss some scenarios.
Again, since the data returned by the queries is pretty small, do the benefit of caching only some query worth the risk and the effort needed to choose them?

Reply:
Would be good to test, what's actually executed often

I think the following are the most important cases:
* lot's of (different) device metadata updates on the registry endpoint
* sending push messages (and the implicit calls to the Flat Metrics)

Than we have "management" API calls, from the UI (E.g. that loads the Application(s), with their Variants, with their Device installations.

Reply:
@matzew : ok, I'm still loading the database to have some GB of data using the ups test suite form @josemigallas. While doing that, I'm checking what queries gets executed.

Reply:
## Creating 1000 APPS, 1 VARIANT and 1 TOKEN
* 500 different SELECTS on ANDROID_VARIANT are executed. Each of them is executed 2 times with the same parameters.
* 1000 unique queries on INSTALLATION
* 3.998 unique queries on VARIANT
* 1.998 unique queries on PUSH_APPLICATION
For a TOTAL of 7996 query.

## Creating 1 APP, 1000 VARIANTS and 0 TOKENS
* 3.000 selects on VARIANT are executed. 1 query is executed 2.000 times always with the same parameters
* 1.945 queries on PUSH_APPLICATION. 1 is executed 1000 times, another is executed 945 times, both with exactly the same parameters

## Creating 1 APP, 1 VARIANT and 1000 TOKENS
* 1000 queries on INSTALLATION
* 1000 queries on ANDROID_VARIANT. It is the same query, executed 1000 times with the same parameters
* 2 queries on PUSH_APPLICATION. The same query is executed 2 times.
* 1002 queries on VARIANT. Here the same exact query is executed 1000 times joining all the variant types.

I currently have the exact query executed on MySQL. Going to match that with the `JPA` code used by UPS.

Reply:
@matzew @gunnarmorling Will update with the matched JPA query and with the data from the other scenario.